### PR TITLE
Allow Decimal values  in String format  to be extracted as double from ResultSet

### PR DESCRIPTION
**Pull request by jetsasank**
This will fix issue  : [5165 ](https://github.com/prestosql/presto/issues/) . The change will allow valid string values in the result set to be converted to a Number (Double) data type. 

```
=== RELEASE NOTES ===

General Change
* Allow Decimal values in String format to be extracted as double from ResultSet.

```

### CODE REVIEW

File: presto-jdbc/src/test/java/io/prestosql/jdbc/BaseTestJdbcResultSet.java [hunk 1]
```diff
@@ -92,6 +92,10 @@ public void testDecimal()
         try (ConnectedStatement connectedStatement = newStatement()) {
             checkRepresentation(connectedStatement.getStatement(), "0.1", Types.DECIMAL, new BigDecimal("0.1"));
             checkRepresentation(connectedStatement.getStatement(), "DECIMAL '0.1'", Types.DECIMAL, new BigDecimal("0.1"));
```
Comment (findepi) [inline presto-jdbc/src/test/java/io/prestosql/jdbc/BaseTestJdbcResultSet.java:94]: add tests for getLong

also, add a test for a value that cannot be represented exact as a double:

```
long outsideOfDoubleExactRange = 9223372036854775774L;
//noinspection ConstantConditions
verify((long) (double) outsideOfDoubleExactRange - outsideOfDoubleExactRange == 33, "outsideOfDoubleExactRange should not be exact-representable as a double");
checkRepresentation(connectedStatement.getStatement(), format("DECIMAL '%s'", outsideOfDoubleExactRange), Types.DECIMAL,
        (rs, column) -> {
            assertEquals(rs.getObject(column), new BigDecimal("9223372036854775774"));
            assertEquals(rs.getBigDecimal(column), new BigDecimal("9223372036854775774"));

            assertEquals(rs.getLong(column), 9223372036854775774L);

            assertEquals(rs.getDouble(column), 9.223372036854776E18);

            assertEquals(rs.getString(column), "9223372036854775774");
        });
```

File: presto-jdbc/src/main/java/io/prestosql/jdbc/AbstractPrestoResultSet.java [hunk 2]
```diff
@@ -1750,6 +1750,9 @@ private static Number toNumber(Object value)
         if (value instanceof Boolean) {
             return ((Boolean) value) ? 1 : 0;
         }
+        if (value instanceof String) {
+            return Double.valueOf((String) value);
```
[the remark this answers is not in the document]
Comment (jetsasank) [reply presto-jdbc/src/main/java/io/prestosql/jdbc/AbstractPrestoResultSet.java]: Good catch, I dint think about precision here. Fixed the code to instead use BigDecimal.

File: presto-jdbc/src/test/java/io/prestosql/jdbc/BaseTestJdbcResultSet.java (hunk 1 above)
Comment (jetsasank) [reply presto-jdbc/src/test/java/io/prestosql/jdbc/BaseTestJdbcResultSet.java:95]: Added this to the unit tests.

**Comment by findepi** [comment]
Thanks for working on this!

Similarly to https://github.com/prestosql/presto/pull/5335#issuecomment-700791964 --

For example, should result set for a query
```
SELECT '1209'; -- a varchar that happens to look like a number
```
return a date upon `ResultSet.getDouble`, `getBigDecimal`, `getLong` etc call?

What do other databases do? PostgreSQL and Oracle in particular?

**Comment by jetsasank** [comment]
> Thanks for working on this!
> 
> Similarly to [#5335 (comment)](https://github.com/prestosql/presto/pull/5335#issuecomment-700791964) --
> 
> For example, should result set for a query
> 
> ```
> SELECT '1209'; -- a varchar that happens to look like a number
> ```
> 
> return a date upon `ResultSet.getDouble`, `getBigDecimal`, `getLong` etc call?
> 
> What do other databases do? PostgreSQL and Oracle in particular?

I was able to verify Postgres and it converts such string into a number. 

```
import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.util.Properties;

class Solution
{
    public static void main(String[] args) throws Exception
    {
        final Properties properties = new Properties();
        properties.put("user", "test");

        Connection connection = DriverManager.getConnection("jdbc:postgresql://localhost/test", properties);

        PreparedStatement ps = connection.prepareStatement("SELECT '1209'");
        ResultSet rs = ps.executeQuery();

        while (rs.next())
        {
            System.out.println(rs.getDouble(1));
            System.out.println(rs.getLong(1));
            System.out.println(rs.getBigDecimal(1));
        }
    }
}
```

Result 

```
1209.0
1209
1209

```

I wasnt able to verify Oracle locally, but based on the docs I've was able to read online. Oracle does something similar too. The code I have here, does similar handling of string for ```getDouble``` and ```getLong```, I might have to handle null and empty string cases. I will add some tests for those edge cases.

**Comment by findepi** [comment]
> I wasnt able to verify Oracle locally

You can start an instance with
```
docker run -p 1521:1521 --rm -i  --name oracle wnameless/oracle-xe-11g-r2
```
(give it some time to start; it prints nothing when ready)


and connect to it with 
```
public static void main(String[] args)
        throws SQLException
{
    try (Connection connection = DriverManager.getConnection("jdbc:oracle:thin:@localhost:1521:xe", "system", "oracle")) {
        try (PreparedStatement statement = connection.prepareStatement("SELECT 1 FROM dual");
                ResultSet resultSet = statement.executeQuery()) {
            verify(resultSet.next());
            System.out.println("resultSet.getInt(1) = " + resultSet.getInt(1));
            verify(!resultSet.next());
        }
    }
}
```

(i put that code in presto-oracle's module to have the Oracle JDBC jar on the classpath)

**Comment by jetsasank** [comment]
> > I wasnt able to verify Oracle locally
> 
> You can start an instance with
> 
> ```
> docker run -p 1521:1521 --rm -i  --name oracle wnameless/oracle-xe-11g-r2
> ```
> 
> (give it some time to start; it prints nothing when ready)
> 
> and connect to it with
> 
> ```
> public static void main(String[] args)
>         throws SQLException
> {
>     try (Connection connection = DriverManager.getConnection("jdbc:oracle:thin:@localhost:1521:xe", "system", "oracle")) {
>         try (PreparedStatement statement = connection.prepareStatement("SELECT 1 FROM dual");
>                 ResultSet resultSet = statement.executeQuery()) {
>             verify(resultSet.next());
>             System.out.println("resultSet.getInt(1) = " + resultSet.getInt(1));
>             verify(!resultSet.next());
>         }
>     }
> }
> ```
> 
> (i put that code in presto-oracle's module to have the Oracle JDBC jar on the classpath)

I was able to test oracle DB and the results are same as Postgres.

```
package io.prestosql.plugin.oracle;

import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.PreparedStatement;
import java.sql.ResultSet;

public class Solution
{
    public static void main(String[] args)
        throws Exception
    {
        Connection connection = DriverManager.getConnection("jdbc:oracle:thin:@localhost:1521:xe", "system", "oracle");
        PreparedStatement ps = connection.prepareStatement("SELECT '1209' from dual");
        ResultSet rs = ps.executeQuery();
        while (rs.next())
        {
            System.out.println(rs.getDouble(1));
            System.out.println(rs.getLong(1));
            System.out.println(rs.getBigDecimal(1));
        }
    }
}
```

```
1209.0
1209
1209
```

File: presto-jdbc/src/main/java/io/prestosql/jdbc/AbstractPrestoResultSet.java [hunk 3]
```diff
@@ -1750,6 +1750,9 @@ private static Number toNumber(Object value)
         if (value instanceof Boolean) {
             return ((Boolean) value) ? 1 : 0;
         }
+        if (value instanceof String) {
+            return new BigDecimal((String) value);
```
Comment (findepi) [inline presto-jdbc/src/main/java/io/prestosql/jdbc/AbstractPrestoResultSet.java:1754]: Add a test case where `rs.get{Long,Double,BigDecimal,..}` are called on a value that is not a number:

- `''`
- `'123a'`

Here we probably will need to catch the exception coming from BigDecimal constructor so that we reach `throw new SQLException("Value is not a number:` line

Comment (jetsasank) [reply presto-jdbc/src/main/java/io/prestosql/jdbc/AbstractPrestoResultSet.java:1754]: Sure.

**Comment by jetsasank** [comment]
@findepi  I had issues with my repo. I recreated the pull request with the discussed changes : https://github.com/prestosql/presto/pull/5509 . Closing this pull request. Sorry for the inconvenience
